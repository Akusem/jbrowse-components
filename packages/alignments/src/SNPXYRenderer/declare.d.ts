declare module 'color'
declare module '@gmod/jbrowse-core/WiggleSNPAbstraction/BaseRenderer'
declare module '@gmod/jbrowse-core/WiggleSNPAbstraction/util'
