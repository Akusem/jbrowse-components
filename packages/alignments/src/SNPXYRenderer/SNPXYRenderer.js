import { readConfObject } from '@gmod/jbrowse-core/configuration'
import { featureSpanPx } from '@gmod/jbrowse-core/util'
import Color from 'color'
import SNPBaseRenderer from '../SNPBaseRenderer'
import { getOrigin, getScale } from '../util'

export default class SNPXYRenderer extends SNPBaseRenderer {
  draw(ctx, props) {
    console.log('props', props)
    const {
      features,
      region,
      bpPerPx,
      scaleOpts,
      height,
      config,
      horizontallyFlipped,
    } = props
    // const pivotValue = readConfObject(config, 'bicolorPivotValue')
    // const negColor = readConfObject(config, 'negColor')
    // const posColor = readConfObject(config, 'posColor')
    // const filled = readConfObject(config, 'filled')
    // const clipColor = readConfObject(config, 'clipColor')
    // const highlightColor = readConfObject(config, 'highlightColor')
    // const summaryScoreMode = readConfObject(config, 'summaryScoreMode')
    // const scale = getScale({ ...scaleOpts, range: [0, height] })
    // const originY = getOrigin(scaleOpts.scaleType)
    // const [niceMin, niceMax] = scale.domain()
    // const toY = rawscore => height - scale(rawscore)
    // const toHeight = rawscore => toY(originY) - toY(rawscore)
    // let colorCallback
    // if (readConfObject(config, 'color') === '#f0f') {
    //   colorCallback = feature =>
    //     feature.get('score') < pivotValue ? negColor : posColor
    // } else {
    //   colorCallback = feature => readConfObject(config, 'color', [feature])
    // }
    let strand

    const leftBase = props.region.start
    const rightBase = props.region.end
    const scale = 1 / props.bpPerPx
    const widthBp = rightBase - leftBase
    const widthPx = widthBp * scale
    const binMax = Math.ceil((rightBase - leftBase) / binWidth)

    const binWidth = function binWidth() {
      return Math.ceil(props.bpPerPx)
    }

    function binNumber(bp) {
      return Math.floor((bp - leftBase) / binWidth)
    }

    const coverageBins = new Array(binMax)
    coverageBins.forEach(function(value, i) {
      coverageBins[i] = new NestedFrequencyTable()
      if (binWidth === 1) coverageBins[i].snpsCounted = true
    })

    const forEachBin = function forEachBin(start, end, callback) {
      let s = (start - leftBase) / binWidth
      let e = (end - 1 - leftBase) / binWidth
      let sb = Math.floor(s)
      let eb = Math.floor(e)

      if (sb >= binMax || eb < 0) return // does not overlap this block

      // enforce 0 <= bin < binMax
      if (sb < 0) {
        s = 0
        sb = 0
      }
      if (eb >= binMax) {
        eb = binMax - 1
        e = binMax
      }

      // now iterate
      if (sb === eb)
        // if in the same bin, just one call
        callback(sb, e - s)
      else {
        // if in different bins, two or more calls
        callback(sb, sb + 1 - s)
        for (let i = sb + 1; i < eb; i++) callback(i, 1)
        callback(eb, e - eb)
      }
    }

    function getFeatures(feature) {
      // move to stats potentially!!!!
      console.log('feature: ', feature)
      strand =
        { '-1': '-', '1': '+' }[`${feature.get('strand')}`] || 'unstranded'
      if (!this.filter(feature)) return

      // increment start and end partial-overlap bins by proportion of overlap
      forEachBin(feature.get('start'), feature.get('end'), function(
        bin,
        overlap,
      ) {
        coverageBins[bin].getNested('reference').increment(strand, overlap)
      })

      // Calculate SNP coverage
      if (binWidth === 1) {
        const mismatches = this.getMismatches(feature)
        // loops through mismatches and updates coverage variables accordingly.
        for (let i = 0; i < mismatches.length; i++) {
          const mismatch = mismatches[i]
          forEachBin(
            feature.get('start') + mismatch.start,
            feature.get('start') + mismatch.start + mismatch.length,
            function(binNumber, overlap) {
              // Note: we decrement 'reference' so that total of the score is the total coverage
              const bin = coverageBins[binNumber]
              bin.getNested('reference').decrement(strand, overlap)
              let { base } = mismatch
              if (mismatch.type == 'insertion') base = `ins ${base}`
              else if (mismatch.type == 'skip') base = 'skip'
              bin.getNested(base).increment(strand, overlap)
            },
          )
        }
      }
    }
    for (const feature of features.values()) {
      console.log(feature)
      ctx.fillRect(0, 100, 100, 100)
    }
    /** ****** IN PROGRESS ***************/
    //   const [leftPx, rightPx] = featureSpanPx(
    //     feature,
    //     region,
    //     bpPerPx,
    //     horizontallyFlipped,
    //   )
    //   let score = feature.get('score')
    //   const maxr = feature.get('maxScore')
    //   const minr = feature.get('minScore')
    //   const lowClipping = score < niceMin
    //   const highClipping = score > niceMax
    //   const w = rightPx - leftPx + 0.3 // fudge factor for subpixel rendering
    //   const c = colorCallback(feature)
    //   if (summaryScoreMode === 'max') {
    //     score = maxr === undefined ? score : maxr
    //     ctx.fillStyle = c
    //     ctx.fillRect(leftPx, toY(score), w, filled ? toHeight(score) : 1)
    //   } else if (summaryScoreMode === 'min') {
    //     score = minr === undefined ? score : minr
    //     ctx.fillStyle = c
    //     ctx.fillRect(leftPx, toY(score), w, filled ? toHeight(score) : 1)
    //   } else if (summaryScoreMode === 'whiskers') {
    //     // max
    //     if (maxr !== undefined) {
    //       ctx.fillStyle = Color(c)
    //         .lighten(0.6)
    //         .toString()
    //       ctx.fillRect(leftPx, toY(maxr), w, filled ? toHeight(maxr) : 1)
    //     }
    //     // normal
    //     ctx.fillStyle = c
    //     ctx.fillRect(leftPx, toY(score), w, filled ? toHeight(score) : 1)
    //     // min
    //     if (minr !== undefined) {
    //       ctx.fillStyle = Color(c)
    //         .darken(0.6)
    //         .toString()
    //       ctx.fillRect(leftPx, toY(minr), w, filled ? toHeight(minr) : 1)
    //     }
    //   } else {
    //     ctx.fillStyle = c
    //     ctx.fillRect(leftPx, toY(score), w, filled ? toHeight(score) : 1)
    //   }
    //   if (highClipping) {
    //     ctx.fillStyle = clipColor
    //     ctx.fillRect(leftPx, 0, w, 4)
    //   } else if (lowClipping && scaleOpts.scaleType !== 'log') {
    //     ctx.fillStyle = clipColor
    //     ctx.fillRect(leftPx, height - 4, w, height)
    //   }
    //   if (feature.get('highlighted')) {
    //     ctx.fillStyle = highlightColor
    //     ctx.fillRect(leftPx, 0, w, height)
    //   }
    // }
    // }
  }
}
