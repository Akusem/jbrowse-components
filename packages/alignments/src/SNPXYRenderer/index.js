import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import ConfigSchema from '../configSchema'

export { default as ReactComponent } from '../SNPRendering'
export { default } from './SNPXYRenderer'

export const configSchema = ConfigurationSchema(
  'SNPXYPlotRenderer',
  {},
  { baseConfiguration: ConfigSchema, explicitlyTyped: true },
)
