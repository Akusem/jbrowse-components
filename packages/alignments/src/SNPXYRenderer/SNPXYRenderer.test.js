import SimpleFeature from '@gmod/jbrowse-core/util/simpleFeature'
import SNPXYRenderer, { configSchema, ReactComponent } from '.'

function SNPXYPlotRendererPlugin() {
  return new SNPXYRenderer({
    name: 'SNPXYRenderer',
    ReactComponent,
    configSchema,
  })
}
test('several features', async () => {
  const result = await SNPXYPlotRendererPlugin().makeImageData({
    features: [
      new SimpleFeature({ id: 't1', data: { start: 1, end: 100, score: 1 } }),
      new SimpleFeature({ id: 't2', data: { start: 101, end: 200, score: 2 } }),
    ],
    region: {
      end: 100,
      start: 1,
    },
    scaleOpts: {
      domain: [0, 50],
      scaleType: 'linear',
    },
    config: {},
    bpPerPx: 3,
    highResolutionScaling: 1,
    horizontallyFlipped: false,
    height: 100,
    featureList: [
      { info: [], position: 100 },
      { info: [], position: 200 },
    ],
  })

  expect(result).toMatchSnapshot({
    imageData: expect.any(Object),
  })
})
