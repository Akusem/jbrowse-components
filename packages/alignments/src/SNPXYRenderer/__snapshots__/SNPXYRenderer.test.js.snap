// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`several features 1`] = `
Object {
  "featureList": Array [
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 1,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 2,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 3,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 4,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 5,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 6,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 7,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 8,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 9,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 10,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 11,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 12,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 13,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 14,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 15,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 16,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 17,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 18,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 19,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 20,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 21,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 22,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 23,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 24,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 25,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 26,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 27,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 28,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 29,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 30,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 31,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 32,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 33,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 34,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 35,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 36,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 37,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 38,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 39,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 40,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 41,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 42,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 43,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 44,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 45,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 46,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 47,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 48,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 49,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 50,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 51,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 52,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 53,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 54,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 55,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 56,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 57,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 58,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 59,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 60,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 61,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 62,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 63,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 64,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 65,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 66,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 67,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 68,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 69,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 70,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 71,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 72,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 73,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 74,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 75,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 76,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 77,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 78,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 79,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 80,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 81,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 82,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 83,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 84,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 85,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 86,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 87,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 88,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 89,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 90,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 91,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 92,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 93,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 94,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 95,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 96,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 97,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 98,
    },
    Object {
      "info": Array [
        Object {
          "base": "reference",
          "score": 1,
          "strands": Object {
            "unstranded": 1,
          },
        },
        Object {
          "base": "total",
          "score": 1,
        },
      ],
      "position": 99,
    },
  ],
  "height": 100,
  "imageData": Any<Object>,
  "width": 33,
}
`;
